# MIMIC-iv-CXR 데이터 분석

> 원본 데이터 및 정보는 [여기](https://physionet.org/content/mimic-cxr/2.1.0/)를 참고하세요.

## 데이터 요약

MIMIC 흉부 X선(MIMIC-CXR) 데이터베이스 v2.1.0은 DICOM 형식의 흉부 방사선 사진과 자유 텍스트 방사선 보고서를 포함한 대규모 공개 데이터 세트입니다. 이 데이터 세트에는 매사추세츠주 보스턴에 있는 Beth Israel Deaconess 의료 센터에서 수행되었으며, 1996년 미국 건강보험 양도 및 책임법(HIPAA)의 세이프 하버(Safe Harbor) 요건을 충족하기 위해 익명화되었습니다.

## 데이터 종류

MIMIC-CXR은 세 가지 데이터 형식이 혼합된 형태입니다.
- 전자 건강 기록 데이터(EHR)
- 이미지(흉부 방사선 사진)
- 자연어(자유 텍스트 보고서)
이 세 가지 양식은 거의 독립적으로 처리되었으며, 데이터베이스를 구축하기 위해 통합되었습니다.

## 레이블링 현황 확인

| 데이터 | 해석 |
|---|---|
|1.0|Positive (예: 환자의 흉부 X-ray에서 Pneumonia(폐렴) 소견이 명확히 있음)|
|-1.0|Uncertain (불확실) 또는 Ambiguous (애매함) (예: 영상에서 폐렴 소견이 있지만 확신할 수 없거나 애매한 경우)|
|NaN|라벨이 부여되지 않은 경우, 즉 판독되지 않았거나 해당 질병에 대한 언급 자체가 없었을 때|

In [3]:
import pandas as pd

chexpert = pd.read_csv('./data/mimic-cxr/mimic-cxr-2.0.0-chexpert.csv.gz', compression='gzip')
chexpert

subject_id  study_id  Atelectasis  Cardiomegaly  Consolidation  Edema  \
0         10000032  50414267          NaN           NaN            NaN    NaN   
1         10000032  53189527          NaN           NaN            NaN    NaN   
2         10000032  53911762          NaN           NaN            NaN    NaN   
3         10000032  56699142          NaN           NaN            NaN    NaN   
4         10000764  57375967          NaN           NaN            1.0    NaN   
...            ...       ...          ...           ...            ...    ...   
227822    19999442  58708861          NaN           NaN            NaN    NaN   
227823    19999733  57132437          NaN           NaN            NaN    NaN   
227824    19999987  55368167          1.0          -1.0            NaN    NaN   
227825    19999987  58621812          1.0           NaN            NaN    NaN   
227826    19999987  58971208          1.0           NaN            NaN    NaN   

        Enlarged Cardiomediastinum  Fracture  Lung Lesion  Lung Opacity  \
0                              NaN       NaN          NaN           NaN   
1                              NaN       NaN          NaN           NaN   
2                              NaN       NaN          NaN           NaN   
3                              NaN       NaN          NaN           NaN   
4                              NaN       NaN          NaN           NaN   
...                            ...       ...          ...           ...   
227822                         NaN       NaN          NaN           NaN   
227823                         NaN       NaN          NaN           NaN   
227824                         NaN       NaN          0.0           NaN   
227825                         NaN       NaN          NaN           NaN   
227826                         NaN       NaN          NaN           NaN   

        No Finding  Pleural Effusion  Pleural Other  Pneumonia  Pneumothorax  \
0              1.0               NaN            NaN        NaN           NaN   
1              1.0               NaN            NaN        NaN           NaN   
2              1.0               NaN            NaN        NaN           NaN   
3              1.0               NaN            NaN        NaN           NaN   
4              NaN               NaN            NaN       -1.0           NaN   
...            ...               ...            ...        ...           ...   
227822         1.0               NaN            NaN        NaN           NaN   
227823         1.0               NaN            NaN        NaN           NaN   
227824         NaN               0.0            NaN        NaN           0.0   
227825         NaN               NaN            NaN        NaN           NaN   
227826         NaN               NaN            NaN        NaN           NaN   

        Support Devices  
0                   NaN  
1                   NaN  
2                   NaN  
3                   NaN  
4                   NaN  
...                 ...  
227822              1.0  
227823              NaN  
227824              NaN  
227825              1.0  
227826              NaN  

[227827 rows x 16 columns]

In [24]:
only_normal = chexpert[(chexpert['No Finding'] == 1.0) & (chexpert['Support Devices'] != 1.0) & (chexpert['Support Devices'] != -1.0) & (chexpert['Pneumonia'] == 0.0)].copy()
only_normal

subject_id  study_id  Atelectasis  Cardiomegaly  Consolidation  Edema  \
15        10000980  54577367          NaN           NaN            NaN    NaN   
34        10001401  58747570          NaN           NaN            NaN    0.0   
52        10001884  55893591          NaN           NaN            NaN    NaN   
114       10002559  59849480          NaN           NaN            NaN    0.0   
121       10003019  53098685          NaN           NaN            NaN    0.0   
...            ...       ...          ...           ...            ...    ...   
227670    19996786  52281280          NaN           NaN            NaN    0.0   
227674    19997087  56876385          NaN           0.0            NaN    0.0   
227708    19997367  52524351          NaN           NaN            NaN    NaN   
227796    19998770  51149538          NaN           NaN            NaN    0.0   
227811    19999270  55274188          NaN           0.0            NaN    0.0   

        Enlarged Cardiomediastinum  Fracture  Lung Lesion  Lung Opacity  \
15                             NaN       NaN          NaN           NaN   
34                             NaN       NaN          NaN           NaN   
52                             NaN       NaN          NaN           NaN   
114                            NaN       NaN          NaN           NaN   
121                            NaN       NaN          NaN           NaN   
...                            ...       ...          ...           ...   
227670                         NaN       NaN          NaN           NaN   
227674                         NaN       NaN          NaN           NaN   
227708                         NaN       NaN          NaN           NaN   
227796                         NaN       NaN          NaN           NaN   
227811                         NaN       NaN          NaN           NaN   

        No Finding  Pleural Effusion  Pleural Other  Pneumonia  Pneumothorax  \
15             1.0               NaN            NaN        0.0           NaN   
34             1.0               NaN            NaN        0.0           NaN   
52             1.0               NaN            NaN        0.0           NaN   
114            1.0               0.0            NaN        0.0           NaN   
121            1.0               0.0            NaN        0.0           NaN   
...            ...               ...            ...        ...           ...   
227670         1.0               0.0            NaN        0.0           NaN   
227674         1.0               0.0            NaN        0.0           NaN   
227708         1.0               NaN            NaN        0.0           NaN   
227796         1.0               0.0            NaN        0.0           NaN   
227811         1.0               0.0            NaN        0.0           NaN   

        Support Devices  
15                  NaN  
34                  NaN  
52                  NaN  
114                 NaN  
121                 NaN  
...                 ...  
227670              NaN  
227674              NaN  
227708              NaN  
227796              NaN  
227811              NaN  

[8507 rows x 16 columns]

In [18]:
result = {
    col: only_normal[col].unique()
    for col in only_normal.columns[2:]
}
result

{'Atelectasis': array([nan,  0.]),
 'Cardiomegaly': array([nan,  0.]),
 'Consolidation': array([nan,  0.]),
 'Edema': array([nan,  0.]),
 'Enlarged Cardiomediastinum': array([nan,  0.]),
 'Fracture': array([nan,  0.]),
 'Lung Lesion': array([nan,  0.]),
 'Lung Opacity': array([nan,  0.]),
 'No Finding': array([1.]),
 'Pleural Effusion': array([nan,  0.]),
 'Pleural Other': array([nan,  0.]),
 'Pneumonia': array([nan,  0.]),
 'Pneumothorax': array([nan,  0.]),
 'Support Devices': array([nan,  0.])}

In [41]:
chexpert['Pneumonia'].unique()

array([nan, -1.,  1.,  0.])

In [15]:
print('총 건수: {:,}'.format(len(chexpert['study_id'].unique())))

총 건수: 227,827


| 질병명                      | 값이 1인 건수 | 값이 1이 아닌 건수 |
|---------------------------|--------------:|--------------------:|
| 무기폐 (Atelectasis)         |        45,808 |             182,019 |
| 심비대 (Cardiomegaly)        |        44,845 |             182,982 |
| 경화 (Consolidation)         |        10,778 |             217,049 |
| 부종 (Edema)                 |        27,018 |             200,809 |
| 심종격 확장 (Enlarged Cardiomediastinum) |     7,179 |             220,648 |
| 골절 (Fracture)              |         4,390 |             223,437 |
| 폐 병변 (Lung Lesion)        |         6,284 |             221,543 |
| 폐 혼탁 (Lung Opacity)       |        51,525 |             176,302 |
| 이상 없음 (No Finding)       |        75,455 |             152,372 |
| 흉막 삼출 (Pleural Effusion) |        54,300 |             173,527 |
| 기타 흉막 질환 (Pleural Other) |       2,011 |             225,816 |
| 폐렴 (Pneumonia)             |        16,556 |             211,271 |
| 기흉 (Pneumothorax)          |        10,358 |             217,469 |
| 인공장치 (Support Devices)   |        66,558 |             161,269 |


**폐렴과 관련된 질병**
- 1차 관련 (가장 직접적):
  - Consolidation
  - Lung Opacity
  - Pleural Effusion
- 2차 관련 (보조적 혹은 감별 필요한 소견):
  - Atelectasis
  - Edema
  - Lung Lesion

In [68]:
# 컬럼별로 1과 1이 아닌 값의 개수 세기
result = []
for col in chexpert.columns[2:]:
    count_nan = chexpert[col].isna().sum()
    count_uncertain = (chexpert[col] == -1).sum()
    count_1 = (chexpert[col] == 1).sum()
    count_0 = (chexpert[col] == 0).sum()
    result.append({'Column': col, 'Count_1': count_1, 'Count_0': count_0, 'Count_uncertain': count_uncertain, 'Count_NaN': count_nan})

# 결과를 DataFrame으로 출력
summary_df = pd.DataFrame(result)
print(summary_df)

                        Column  Count_1  Count_0  Count_uncertain  Count_NaN
0                  Atelectasis    45808     1531            10327     170161
1                 Cardiomegaly    44845    15911             6043     161028
2                Consolidation    10778     7967             4331     204751
3                        Edema    27018    25641            13174     161994
4   Enlarged Cardiomediastinum     7179     5283             9375     205990
5                     Fracture     4390      886              555     221996
6                  Lung Lesion     6284      862             1141     219540
7                 Lung Opacity    51525     3069             3831     169402
8                   No Finding    75455        0                0     152372
9             Pleural Effusion    54300    27158             5814     140555
10               Pleural Other     2011      126              765     224925
11                   Pneumonia    16556    24338            18291     168642

- 정상 건수: 75,455
- 폐렴(Pneumonia) 건수: 16,556
- 폐렴(Pneumonia) 아닌 비정상 건수: 135816  (= 227827 - 75455 - 16556)

## 폐렴(Pneumonia) 데이터 다운로드

In [ ]:
# no finding
24338+8851

33189

In [69]:
# finding
6012 + 16556

22568

In [ ]:
# Pneumonia = 1 일 때 다른 컬럼이 1과 1이 아닌 값의 개수 세기
pneumonia_cases = chexpert[chexpert['Pneumonia']==1.0]
pneumonia_cases = pneumonia_cases.drop(columns='Pneumonia')

# 컬럼별로 1과 1이 아닌 값의 개수 세기
result = []
for col in pneumonia_cases.columns[2:]:
    count_1 = (pneumonia_cases[col] == 1).sum()
    count_not_1 = (pneumonia_cases[col] != 1).sum()
    result.append({'Column': col, 'Count_1': count_1, 'Count_not_1': count_not_1})

# 결과를 DataFrame으로 출력
summary_df = pd.DataFrame(result)
print(summary_df)

count = (pneumonia_cases[:] != 0).all()
count

                        Column  Count_1  Count_not_1
0                  Atelectasis     3672        12884
1                 Cardiomegaly     3507        13049
2                Consolidation     2268        14288
3                        Edema     3052        13504
4   Enlarged Cardiomediastinum      546        16010
5                     Fracture      180        16376
6                  Lung Lesion      703        15853
7                 Lung Opacity     8397         8159
8                   No Finding        0        16556
9             Pleural Effusion     5007        11549
10               Pleural Other      206        16350
11                Pneumothorax      280        16276
12             Support Devices     4543        12013


In [57]:
is_only_pneumonia = is_only_pneumonia = (pneumonia_cases != 1.0).all(axis=1)
count = is_only_pneumonia.sum()
count


np.int64(3045)

In [55]:
count = 0

for _, row in chexpert.iterrows():
    if row['Pneumonia'] == 1.0 and (row.drop('Pneumonia')!=0).all():
        count += 1

print('폐렴(Pneumonia)이면서 다른 이상이 없는 총 건수: {:,}'.format(count))

폐렴(Pneumonia)이면서 다른 이상이 없는 총 건수: 10,360


In [ ]:
meta = pd.read_csv('mimic-cxr-2.0.0-metadata.csv.gz', compression='gzip')
meta

dicom_id  subject_id  study_id  \
0       02aa804e-bde0afdd-112c0b34-7bc16630-4e384014    10000032  50414267   
1       174413ec-4ec4c1f7-34ea26b7-c5f994f8-79ef1962    10000032  50414267   
2       2a2277a9-b0ded155-c0de8eb9-c124d10e-82c5caab    10000032  53189527   
3       e084de3b-be89b11e-20fe3f9f-9c8d8dfe-4cfd202c    10000032  53189527   
4       68b5c4b1-227d0485-9cc38c3f-7b84ab51-4b472714    10000032  53911762   
...                                              ...         ...       ...   
377105  428e2c18-5721d8f3-35a05001-36f3d080-9053b83c    19999733  57132437   
377106  58c403aa-35ff8bd9-73e39f54-8dc9cc5d-e0ec3fa9    19999733  57132437   
377107  58766883-376a15ce-3b323a28-6af950a0-16b793bd    19999987  55368167   
377108  7ba273af-3d290f8d-e28d0ab4-484b7a86-7fc12b08    19999987  58621812   
377109  1a1fe7e3-cbac5d93-b339aeda-86bb86b5-4f31e82e    19999987  58971208   

       PerformedProcedureStepDescription ViewPosition  Rows  Columns  \
0                     CHEST (PA AND LAT)           PA  3056     2544   
1                     CHEST (PA AND LAT)      LATERAL  3056     2544   
2                     CHEST (PA AND LAT)           PA  3056     2544   
3                     CHEST (PA AND LAT)      LATERAL  3056     2544   
4                    CHEST (PORTABLE AP)           AP  2705     2539   
...                                  ...          ...   ...      ...   
377105                CHEST (PA AND LAT)           PA  3056     2544   
377106                CHEST (PA AND LAT)      LATERAL  3056     2544   
377107               CHEST (PORTABLE AP)           AP  2544     3056   
377108               CHEST (PORTABLE AP)           AP  3056     2544   
377109               CHEST (PORTABLE AP)           AP  3056     2544   

        StudyDate   StudyTime ProcedureCodeSequence_CodeMeaning  \
0        21800506  213014.531                CHEST (PA AND LAT)   
1        21800506  213014.531                CHEST (PA AND LAT)   
2        21800626  165500.312                CHEST (PA AND LAT)   
3        21800626  165500.312                CHEST (PA AND LAT)   
4        21800723   80556.875               CHEST (PORTABLE AP)   
...           ...         ...                               ...   
377105   21520708  224550.171                CHEST (PA AND LAT)   
377106   21520708  224550.171                CHEST (PA AND LAT)   
377107   21451104   51448.218               CHEST (PORTABLE AP)   
377108   21451102  202809.234               CHEST (PORTABLE AP)   
377109   21451103   50507.625               CHEST (PORTABLE AP)   

       ViewCodeSequence_CodeMeaning PatientOrientationCodeSequence_CodeMeaning  
0                  postero-anterior                                      Erect  
1                           lateral                                      Erect  
2                  postero-anterior                                      Erect  
3                           lateral                                      Erect  
4                  antero-posterior                                        NaN  
...                             ...                                        ...  
377105             postero-anterior                                      Erect  
377106                      lateral                                      Erect  
377107             antero-posterior                                      Erect  
377108             antero-posterior                                      Erect  
377109             antero-posterior                                  Recumbent  

[377110 rows x 12 columns]

In [24]:
print('{:,}'.format(len(meta['subject_id'].unique())))
print('{:,}'.format(len(meta['study_id'].unique())))
print('{:,}'.format(meta['dicom_id'].count()))

65,379
227,835
377,110


In [25]:
negbio = pd.read_csv('mimic-cxr-2.0.0-negbio.csv.gz', compression='gzip')
negbio

subject_id  study_id  Atelectasis  Cardiomegaly  Consolidation  Edema  \
0         10000032  50414267          NaN           NaN            NaN    NaN   
1         10000032  53189527          NaN           NaN            NaN    NaN   
2         10000032  53911762          NaN           NaN            NaN    NaN   
3         10000032  56699142          NaN           NaN            NaN    NaN   
4         10000764  57375967          NaN           NaN            1.0    NaN   
...            ...       ...          ...           ...            ...    ...   
227822    19999442  58708861          NaN           NaN            NaN    NaN   
227823    19999733  57132437          NaN           NaN            NaN    NaN   
227824    19999987  55368167          1.0          -1.0            NaN    NaN   
227825    19999987  58621812          1.0           NaN            NaN    NaN   
227826    19999987  58971208          1.0           NaN            NaN    NaN   

        Enlarged Cardiomediastinum  Fracture  Lung Lesion  Lung Opacity  \
0                              NaN       NaN          NaN           NaN   
1                              NaN       NaN          NaN           NaN   
2                              NaN       NaN          NaN           NaN   
3                              NaN       NaN          NaN           NaN   
4                              NaN       NaN          NaN           NaN   
...                            ...       ...          ...           ...   
227822                         NaN       NaN          NaN           NaN   
227823                         NaN       NaN          NaN           NaN   
227824                         NaN       NaN          0.0           NaN   
227825                         NaN       NaN          NaN           NaN   
227826                         NaN       NaN          NaN           NaN   

        No Finding  Pleural Effusion  Pleural Other  Pneumonia  Pneumothorax  \
0              1.0               NaN            NaN        NaN           NaN   
1              1.0               NaN            NaN        NaN           NaN   
2              1.0               NaN            NaN        NaN           NaN   
3              1.0               NaN            NaN        NaN           NaN   
4              NaN               NaN            NaN       -1.0           NaN   
...            ...               ...            ...        ...           ...   
227822         1.0               NaN            NaN        NaN           NaN   
227823         1.0               NaN            NaN        NaN           NaN   
227824         NaN               0.0            NaN        NaN           0.0   
227825         NaN               NaN            NaN        NaN           NaN   
227826         NaN               NaN            NaN        NaN           NaN   

        Support Devices  
0                   NaN  
1                   NaN  
2                   NaN  
3                   NaN  
4                   NaN  
...                 ...  
227822              1.0  
227823              NaN  
227824              NaN  
227825              1.0  
227826              NaN  

[227827 rows x 16 columns]

In [26]:
split = pd.read_csv('mimic-cxr-2.0.0-split.csv.gz', compression='gzip')
split

dicom_id  study_id  subject_id  \
0       02aa804e-bde0afdd-112c0b34-7bc16630-4e384014  50414267    10000032   
1       174413ec-4ec4c1f7-34ea26b7-c5f994f8-79ef1962  50414267    10000032   
2       2a2277a9-b0ded155-c0de8eb9-c124d10e-82c5caab  53189527    10000032   
3       e084de3b-be89b11e-20fe3f9f-9c8d8dfe-4cfd202c  53189527    10000032   
4       68b5c4b1-227d0485-9cc38c3f-7b84ab51-4b472714  53911762    10000032   
...                                              ...       ...         ...   
377105  428e2c18-5721d8f3-35a05001-36f3d080-9053b83c  57132437    19999733   
377106  58c403aa-35ff8bd9-73e39f54-8dc9cc5d-e0ec3fa9  57132437    19999733   
377107  58766883-376a15ce-3b323a28-6af950a0-16b793bd  55368167    19999987   
377108  7ba273af-3d290f8d-e28d0ab4-484b7a86-7fc12b08  58621812    19999987   
377109  1a1fe7e3-cbac5d93-b339aeda-86bb86b5-4f31e82e  58971208    19999987   

        split  
0       train  
1       train  
2       train  
3       train  
4       train  
...       ...  
377105  train  
377106  train  
377107  train  
377108  train  
377109  train  

[377110 rows x 4 columns]

# RSNA Pneumonia Detection Challenge 데이터 분석

In [58]:
import pandas as pd

rsna = pd.read_csv('./data/rsna/stage_2_detailed_class_info.csv')
rsna.head()

patientId                         class
0  0004cfab-14fd-4e49-80ba-63a80b6bddd6  No Lung Opacity / Not Normal
1  00313ee0-9eaa-42f4-b0ab-c148ed3241cd  No Lung Opacity / Not Normal
2  00322d4d-1c29-4943-afc9-b6754be640eb  No Lung Opacity / Not Normal
3  003d8fa0-6bf1-40ed-b54c-ac657f8495c5                        Normal
4  00436515-870c-4b36-a041-de91049b9ab4                  Lung Opacity

In [60]:
class_0 = (rsna['class']=='Normal').sum()
class_1 = (rsna['class']=='Lung Opacity').sum()
class_2 = (rsna['class']=='No Lung Opacity / Not Normal').sum()

print(class_0, class_1, class_2)

8851 9555 11821


In [61]:
rsna

patientId                         class
0      0004cfab-14fd-4e49-80ba-63a80b6bddd6  No Lung Opacity / Not Normal
1      00313ee0-9eaa-42f4-b0ab-c148ed3241cd  No Lung Opacity / Not Normal
2      00322d4d-1c29-4943-afc9-b6754be640eb  No Lung Opacity / Not Normal
3      003d8fa0-6bf1-40ed-b54c-ac657f8495c5                        Normal
4      00436515-870c-4b36-a041-de91049b9ab4                  Lung Opacity
...                                     ...                           ...
30222  c1ec14ff-f6d7-4b38-b0cb-fe07041cbdc8                  Lung Opacity
30223  c1edf42b-5958-47ff-a1e7-4f23d99583ba                        Normal
30224  c1f6b555-2eb1-4231-98f6-50a963976431                        Normal
30225  c1f7889a-9ea9-4acb-b64c-b737c929599a                  Lung Opacity
30226  c1f7889a-9ea9-4acb-b64c-b737c929599a                  Lung Opacity

[30227 rows x 2 columns]